In [12]:
#Import libraries below

In [57]:
import numpy as np 
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [30]:
#Import the data and look at it - csv file for shark attacks

In [19]:
data = pd.read_csv("../desktop/attacks.csv",encoding = "ISO-8859-1")


In [32]:
data.shape
print("the dataframe has 25,723 rows and 24 columns")

the dataframe has 25,723 rows and 24 columns


In [59]:
data.head(10)

Case Number         Date    Year        Type    Country             Area  \
0    2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1    2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2    2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3    2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4    2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   
5  2018.06.03.b  03-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
6  2018.06.03.a  03-Jun-2018  2018.0  Unprovoked     BRAZIL       Pernambuco   
7    2018.05.27  27-May-2018  2018.0  Unprovoked        USA          Florida   
8  2018.05.26.b  26-May-2018  2018.0  Unprovoked        USA          Florida   
9  2018.05.26.a  26-May-2018  2018.0  Unprovoked        USA          Florida   

                                            Location      Activity  \
0                        Oceanside, San Diego County      Paddling   
1                     St. Simon Island, Glynn County      Standing   
2                                       Habush, Oahu       Surfing   
3                                 Arrawarra Headland       Surfing   
4                                           La Ticla   Free diving   
5                                 Flat Rock, Ballina  Kite surfing   
6                              Piedade Beach, Recife      Swimming   
7  Lighhouse Point Park, Ponce Inlet, Volusia County       Fishing   
8                       Cocoa Beach, Brevard  County       Walking   
9                      Daytona Beach, Volusia County      Standing   

                     Name Sex   ...         Species   \
0             Julie Wolfe    F  ...      White shark   
1         Adyson McNeely     F  ...              NaN   
2             John Denges    M  ...              NaN   
3                    male    M  ...        2 m shark   
4          Gustavo Ramos     M  ...  Tiger shark, 3m   
5                 Chris 
    M  ...              NaN   
6  Jose Ernesto da Silva     M  ...      Tiger shark   
7                    male    M  ...  Lemon shark, 3'   
8               Cody High    M  ...   Bull shark, 6'   
9                    male    M  ...              NaN   

             Investigator or Source                            pdf  \
0                  R. Collier, GSAF           2018.06.25-Wolfe.pdf   
1    K.McMurray, TrackingSharks.com         2018.06.18-McNeely.pdf   
2    K.McMurray, TrackingSharks.com          2018.06.09-Denges.pdf   
3                    B. Myatt, GSAF       2018.06.08-Arrawarra.pdf   
4                         A .Kipper           2018.06.04-Ramos.pdf   
5         Daily Telegraph, 6/4/2018      2018.06.03.b-FlatRock.pdf   
6    Diario de Pernambuco, 6/4/2018       2018.06.03.a-daSilva.pdf   
7   K. McMurray, TrackingSharks.com           2018.05.27-Ponce.pdf   
8    K.McMurray, TrackingSharks.com          2018.05.26.b-High.pdf   
9  K. McMurray, Tracking Sharks.com  2018.05.26.a-DaytonaBeach.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   
5  http://sharkattackfile.net/spreadsheets/pdf_di...   
6  http://sharkattackfile.net/spreadsheets/pdf_di...   
7  http://sharkattackfile.net/spreadsheets/pdf_di...   
8  http://sharkattackfile.net/spreadsheets/pdf_di...   
9  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08 

In [60]:
data.tail(10)

Case Number Date  Year Type Country Area Location Activity Name Sex   \
25713         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25714         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25715         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25716         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25717         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25718         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25719         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25720         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25721         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25722          xx  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

       ... Species  Investigator or Source  pdf href formula href  \
25713  ...      NaN                    NaN  NaN          NaN  NaN   
25714  ...      NaN                    NaN  NaN          NaN  NaN   
25715  ...      NaN                    NaN  NaN          NaN  NaN   
25716  ...      NaN                    NaN  NaN          NaN  NaN   
25717  ...      NaN                    NaN  NaN          NaN  NaN   
25718  ...      NaN                    NaN  NaN          NaN  NaN   
25719  ...      NaN                    NaN  NaN          NaN  NaN   
25720  ...      NaN                    NaN  NaN          NaN  NaN   
25721  ...      NaN                    NaN  NaN          NaN  NaN   
25722  ...      NaN                    NaN  NaN          NaN  NaN   

      Case Number.1 Case Number.2 original order Unnamed: 22 Unnamed: 23  
25713           NaN           NaN            NaN         NaN         NaN  
25714           NaN           NaN            NaN         NaN         NaN  
25715           NaN           NaN            NaN         NaN         NaN  
25716           NaN           NaN            NaN         NaN         NaN  
25717           NaN           NaN            NaN         NaN         NaN  
25718           NaN           NaN            NaN         NaN         NaN  
25719           NaN           NaN            NaN         NaN         NaN  
25720           NaN           NaN            NaN         NaN         NaN  
25721           NaN           NaN            NaN         NaN         NaN  
25722           NaN           NaN            NaN         NaN         NaN  

[10 rows x 24 columns]

In [28]:
print(data.columns)

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')


In [29]:
print(data.keys)

<bound method NDFrame.keys of       Case Number         Date    Year        Type    Country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating        USA   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA   
2      2018.06.09  09-Jun-2018  2018.0     Invalid        USA   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
4      2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO   
...           ...          ...     ...         ...        ...   
25718         NaN          NaN     NaN         NaN        NaN   
25719         NaN          NaN     NaN         NaN        NaN   
25720         NaN          NaN     NaN         NaN        NaN   
25721         NaN          NaN     NaN         NaN        NaN   
25722          xx          NaN     NaN         NaN        NaN   

                  Area                        Location     Activity  \
0           California     Oceanside, San Diego County     Paddling   
1              Georgia  St. Simon Island, Glynn

In [33]:
#Chek repetitions

In [35]:
print("Some rows may be repeated, in order to make sure this does not happen I will use the 'drop.duplicates' function.")

Some columns may be repeated, in order to make sure this does not happen I will use the 'drop.duplicates' function.


In [37]:
actual_data = data.drop_duplicates()
actual_data.shape

(6312, 24)

In [44]:
count_row = data.shape[0]
print(count_row)
count_row2 =actual_data.shape[0]
print(count_row2)

25723
6312


In [46]:
rows_repeated = count_row - count_row2
print(rows_repeated)

19411


In [47]:
print("It was a good idea to check for repeated rows because now we only have,",count_row2,"rows,",rows_repeated,"less than before.")

It was a good idea to check for repeated rows because now we only have, 6312 rows, 19411 less than before.


In [50]:
print(actual_data.isnull())
print("Using the isnull() method I can see that there are some empty values n/a, and now I want to see in which places those values are misssing")

       Case Number   Date   Year   Type  Country   Area  Location  Activity  \
0            False  False  False  False    False  False     False     False   
1            False  False  False  False    False  False     False     False   
2            False  False  False  False    False  False     False     False   
3            False  False  False  False    False  False     False     False   
4            False  False  False  False    False  False     False     False   
...            ...    ...    ...    ...      ...    ...       ...       ...   
6307         False   True   True   True     True   True      True      True   
6308         False   True   True   True     True   True      True      True   
6309         False   True   True   True     True   True      True      True   
8702          True   True   True   True     True   True      True      True   
25722        False   True   True   True     True   True      True      True   

        Name   Sex   ...  Species   Investigator or

In [54]:
print(actual_data.isnull().sum())

Case Number                  2
Date                        10
Year                        12
Type                        14
Country                     60
Area                       465
Location                   550
Activity                   554
Name                       220
Sex                        575
Age                       2841
Injury                      38
Fatal (Y/N)                549
Time                      3364
Species                   2848
Investigator or Source      27
pdf                         10
href formula                11
href                        10
Case Number.1               10
Case Number.2               10
original order               3
Unnamed: 22               6311
Unnamed: 23               6310
dtype: int64


In [56]:
print("Now we can see how many values are missing for each column, and it appears that, Unnamed:23 and Unnamed:24 suffer most from this problem, followed by Time, Species, and Age.")

Now we can see how many values are missing for each column, and it appears that, Unnamed:23 and Unnamed:24 suffer most from this problem, followed by Time, Species, and Age.


There are too many columns describind the date of the attack. So im going to drop 'Case Number.1', 'Case Number.2', and keep the 'Case number' and the 'Date' columns

In [86]:
actual_data2 = actual_data.drop(columns=['Case Number.1', 'Case Number.2'])

In [88]:
actual_data2

Case Number         Date    Year        Type    Country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating        USA   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA   
2      2018.06.09  09-Jun-2018  2018.0     Invalid        USA   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
4      2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO   
...           ...          ...     ...         ...        ...   
6307            0          nan     NaN         NaN        NaN   
6308            0          nan     NaN         NaN        NaN   
6309            0          nan     NaN         NaN        NaN   
8702          NaN          nan     NaN         NaN        NaN   
25722          xx          nan     NaN         NaN        NaN   

                  Area                        Location     Activity  \
0           California     Oceanside, San Diego County     Paddling   
1              Georgia  St. Simon Island, Glynn County     Standing   
2               Hawaii                    Habush, Oahu      Surfing   
3      New South Wales              Arrawarra Headland      Surfing   
4               Colima                        La Ticla  Free diving   
...                ...                             ...          ...   
6307               NaN                             NaN          NaN   
6308               NaN                             NaN          NaN   
6309               NaN                             NaN          NaN   
8702               NaN                             NaN          NaN   
25722              NaN                             NaN          NaN   

                  Name Sex   ... Fatal (Y/N)           Time         Species   \
0          Julie Wolfe    F  ...           N          18h00      White shark   
1      Adyson McNeely     F  ...           N  14h00  -15h00              NaN   
2          John Denges    M  ...           N          07h45              NaN   
3                 male    M  ...           N            NaN        2 m shark   
4       Gustavo Ramos     M  ...           N            NaN  Tiger shark, 3m   
...                ...  ...  ...         ...            ...              ...   
6307               NaN  NaN  ...         NaN            NaN              NaN   
6308               NaN  NaN  ...         NaN            NaN              NaN   
6309               NaN  NaN  ...         NaN            NaN              NaN   
8702               NaN  NaN  ...         NaN            NaN              NaN   
25722              NaN  NaN  ...         NaN            NaN              NaN   

               Investigator or Source                       pdf  \
0                    R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1      K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2      K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3                      B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4                           A .Kipper      2018.06.04-Ramos.pdf   
...                               ...                       ...   
6307                              NaN                       NaN   
6308                              NaN                       NaN   
6309                              NaN                       NaN   
8702                              NaN                       NaN   
25722                             NaN                       NaN   

                                            href formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...   
1      http://sharkattackfile.net/spreadsheets/pdf_di...   
2      http://sharkattackfile.net/spreadsheets/pdf_di...   
3      http://sharkattackfile.net/spreadsheets/pdf_di...   
4      http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                  ...   
6307                                                 NaN   
6308                                                 NaN   
6309                                                 NaN   
8702  

Now we have 22 columns

I believe that all the rows under column unnamed:22 and unnamed:23 have missing values and therefore are not usefull. If i find that to be true I will also drop those columns.


In [89]:
actual_data2['Unnamed: 22'].isnull().sum()


6311

In [90]:
actual_data2['Unnamed: 23'].isnull().sum()


6310

In [91]:
actual_data2.shape

(6312, 22)

In [92]:
actual_data3 = actual_data2.drop(columns=['Unnamed: 22', 'Unnamed: 23'])

In [93]:
actual_data3.shape

(6312, 20)

I have dropped both colums because they did not have any significant data. Now we have 20 columns in our dataframe.

I will also drop the columns that I believe will have no use in my plotting and examination of the data.

In [109]:
final_data = actual_data3.drop(columns=['Case Number', 'Species ', 'Location', 'Area', 'pdf', 'Time', 'Investigator or Source', 'href formula', 'href', 'original order'])

In [110]:
final_data.shape

(6312, 10)

In [111]:
final_data.head()

Date    Year        Type    Country     Activity             Name  \
0  25-Jun-2018  2018.0     Boating        USA     Paddling      Julie Wolfe   
1  18-Jun-2018  2018.0  Unprovoked        USA     Standing  Adyson McNeely    
2  09-Jun-2018  2018.0     Invalid        USA      Surfing      John Denges   
3  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA      Surfing             male   
4  04-Jun-2018  2018.0    Provoked     MEXICO  Free diving   Gustavo Ramos    

  Sex   Age                                             Injury Fatal (Y/N)  
0    F   57  No injury to occupant, outrigger canoe and pad...           N  
1    F   11                         Minor injury to left thigh           N  
2    M   48       Injury to left lower leg from surfboard skeg           N  
3    M  NaN                          Minor injury to lower leg           N  
4    M  NaN  Lacerations to leg & hand shark PROVOKED INCIDENT           N

In [112]:
final_data.tail()

Date  Year Type Country Activity Name Sex   Age Injury Fatal (Y/N)
6307   nan   NaN  NaN     NaN      NaN  NaN  NaN  NaN    NaN         NaN
6308   nan   NaN  NaN     NaN      NaN  NaN  NaN  NaN    NaN         NaN
6309   nan   NaN  NaN     NaN      NaN  NaN  NaN  NaN    NaN         NaN
8702   nan   NaN  NaN     NaN      NaN  NaN  NaN  NaN    NaN         NaN
25722  nan   NaN  NaN     NaN      NaN  NaN  NaN  NaN    NaN         NaN

There are still rows with missing values so i will go ahed and delete those.

In [113]:
final_data.isnull().sum().sort_values(ascending=False)

Age            2841
Sex             575
Activity        554
Fatal (Y/N)     549
Name            220
Country          60
Injury           38
Type             14
Year             12
Date              0
dtype: int64

In [122]:
final_data.dropna()

Date    Year        Type             Country  \
0     25-Jun-2018  2018.0     Boating                 USA   
1     18-Jun-2018  2018.0  Unprovoked                 USA   
2     09-Jun-2018  2018.0     Invalid                 USA   
6     03-Jun-2018  2018.0  Unprovoked              BRAZIL   
7     27-May-2018  2018.0  Unprovoked                 USA   
...           ...     ...         ...                 ...   
6243        1960s     0.0  Unprovoked                IRAQ   
6254  Before 1957     0.0    Provoked                CUBA   
6265    1941-1942     0.0  Unprovoked                IRAQ   
6276  Early 1930s     0.0  Unprovoked              BELIZE   
6301    1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)   

                                               Activity  \
0                                              Paddling   
1                                              Standing   
2                                               Surfing   
6                                              Swimming   
7                                               Fishing   
...                                                 ...   
6243  Swimming in section of river used for washing ...   
6254                   Shark fishing, knocked overboard   
6265                                           Swimming   
6276                                           Standing   
6301                                           Swimming   

                        Name Sex        Age  \
0                Julie Wolfe    F        57   
1            Adyson McNeely     F        11   
2                John Denges    M        48   
6     Jose Ernesto da Silva     M        18   
7                       male    M        52   
...                      ...  ...       ...   
6243                    male    M        16   
6254              Sandrillio    M        50   
6265                    male    M  13 or 14   
6276               a servant    M        16   
6301                    male    M        15   

                                                 Injury Fatal (Y/N)  
0     No injury to occupant, outrigger canoe and pad...           N  
1                            Minor injury to left thigh           N  
2          Injury to left lower leg from surfboard skeg           N  
6                                                 FATAL           Y  
7               Minor injury to foot. PROVOKED INCIDENT           N  
...                                                 ...         ...  
6243         Right leg lacerated & surgically amputated           N  
6254               FATAL, hip bitten  PROVOKED INCIDENT           Y  
6265      FATAL, left leg bitten with severe blood loss           Y  
6276                                              FATAL           Y  
6301  FATAL. "Shark bit him in half, carrying away t...           Y  

[3091 rows x 10 columns]

Now i have my final data. With 3091 rows and 10 columns which will be useful in order to find evidence and correlations in the data.

In [123]:
final_data.to_csv(r'desktop\clean_data.csv')